### PR TITLE
Enable QDWH TPU tests.

### PR DESCRIPTION
**Pull request by tlu7**
TODO for  `testQdwhWithRandomMatrix` indicates TPU failure, whereas both `gpu` and `tpu` are disabled.  Please let me know if I should disable `gpu` if there is any A100 failue.

>  TODO(tianjianlu): Fails on TPU.
  @jtu.skip_on_devices("gpu", "tpu")
  def testQdwhWithRandomMatrix(self, m, n, log_cond):

### CODE REVIEW

**Comment by tlu7** [comment]
@hawkinsp

File: tests/qdwh_test.py [hunk 1]
```diff
@@ -103,30 +106,31 @@ def testQdwhWithUpperTriangularInputAllOnes(self, m, n, log_cond):
...
+      relative_diff_u = _compute_relative_diff(actual_u, expected_u)
+      np.testing.assert_almost_equal(relative_diff_u, 1E-4, decimal=2)
 
     with self.subTest('Test h.'):
-      self.assertAllClose(actual_h, expected_h, rtol=rtol)
+      relative_diff_h = _compute_relative_diff(actual_h, expected_h)
+      np.testing.assert_almost_equal(relative_diff_h, 1E-4, decimal=2)
 
     with self.subTest('Test u.dot(h).'):
       a_round_trip = actual_u.dot(actual_h)
```
Comment (hawkinsp) [inline tests/qdwh_test.py:117]: Is it possible that the TPU precision problems occur because JAX currently defaults to low precision for matmuls on TPU? I think calling `.dot` will do that.

Try using the scope:
```
with jax.default_matmul_precision('float32'):
```

or passing the `precision=` argument to methods like `dot`.

**Comment by tlu7** [comment]
Thanks for the offline discussion. Turned on highest precision globally in qdwh.py and at the matmul in qdwh_test.py; and saw two orders of magnitude higher accuracy in the test.

**Comment by tlu7** [comment]
The test failure is related to the precision check. I am looking at it. 
[https://github.dev/google/jax/blob/b0d334a8811e5fc38a5e875e42d598c06e229c83/tests/lax_numpy_test.py#L5506-L5507](url)

**Comment by tlu7** [comment]
> The test failure is related to the precision check. I am looking at it.
> 
> https://github.com/google/jax/blob/b0d334a8811e5fc38a5e875e42d598c06e229c83/tests/lax_numpy_test.py#L5506

Here are the findings:

The test failure in `lax_numpy_test` is due to `jax.config.update('jax_default_matmul_precision', 'float32')` in qdwh.py.

Removing `jax.config.update('jax_default_matmul_precision', 'float32')` in qdwh.py does not affect the test accuracy (this is most likely due to the function calls are all within the scope of `with jax.default_matmul_precision('float32'):`.

File: tests/qdwh_test.py [hunk 2]
```diff
@@ -85,38 +89,45 @@ def testQdwhUnconvergedAfterMaxNumberIterations(
...
 
     with self.subTest('Test h.'):
-      self.assertAllClose(actual_h, expected_h, rtol=rtol)
+      relative_diff_h = _compute_relative_diff(actual_h, expected_h)
+      np.testing.assert_almost_equal(relative_diff_h, 1E-6, decimal=5)
 
     with self.subTest('Test u.dot(h).'):
-      a_round_trip = actual_u.dot(actual_h)
-      self.assertAllClose(a_round_trip, a, rtol=rtol)
+      with jax.default_matmul_precision('float32'):
```
Comment (hawkinsp) [inline tests/qdwh_test.py:120]: You might find it cleaner to simply define a helper function:
```
_dot = partial(jnp.dot, precision="highest")
```
and use it everywhere in this test file, rather than sticking the scope in many places?

File: tests/qdwh_test.py [hunk 3]
```diff
@@ -85,38 +89,45 @@ def testQdwhUnconvergedAfterMaxNumberIterations(
...
+    a = jnp.triu(jnp.ones((m, n))).astype(_QDWH_TEST_DTYPE)
     u, s, v = jnp.linalg.svd(a, full_matrices=False)
     cond = 10**log_cond
     s = jnp.linspace(cond, 1, min(m, n))
     a = (u * s) @ v
     is_symmetric = _check_symmetry(a)
     max_iterations = 10
-    actual_u, actual_h, _, _ = qdwh.qdwh(a, is_symmetric, max_iterations)
+
+    with jax.default_matmul_precision('float32'):
```
Comment (hawkinsp) [inline tests/qdwh_test.py:104]: I would not add the scope around the call to qdwh. I would add the necessary precision overrides inside the qdwh implementation, since I'm not sure we have any evidence that QDWH works well with low matmul precision.

**Comment by tlu7** [comment]
Thanks for the comments! PTAL.

**Review by hawkinsp — APPROVED** [review]
